v2/api/dbforpostgresql/customizations/flexible_server_extensions.go

```diff
@@ -79,3 +85,55 @@ func secretsToWrite(obj *postgresql.FlexibleServer) ([]*v1.Secret, error) {
 
 	return collector.Values()
 }
+
+var _ extensions.PreReconciliationChecker = &FlexibleServerExtension{}
+
+// If a flexible server does not have a provisioningState in this set, it will reject any attempt to PUT a new state
+// out of hand; so there's no point in even trying. This is true even if the PUT we're doing will have no effect on
+// the state of the cluster.
+// These are all listed lowercase, so we can do a case-insensitive match.
+var nonBlockingFlexibleServerStates = set.Make(
+	"succeeded",
+	"failed",
+	"canceled",
+)
+
+func (ext *FlexibleServerExtension) PreReconcileCheck(
+	_ context.Context,
+	obj genruntime.MetaObject,
+	owner genruntime.MetaObject,
+	_ kubeclient.Client,
+	_ *genericarmclient.GenericClient,
+	_ logr.Logger,
+	_ extensions.PreReconcileCheckFunc,
+) (extensions.PreReconcileCheckResult, error) {
+	// This has to be the current hub storage version. It will need to be updated
+	// if the hub storage version changes.
+	server, ok := obj.(*postgresql.FlexibleServer)
+	if !ok {
+		return extensions.PreReconcileCheckResult{},
+			errors.Errorf("cannot run on unknown resource type %T, expected *postgresql.FlexibleServer", obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not been updated
+	var _ conversion.Hub = server
+
+	if flexibleServerStateBlocksReconciliation(server) {
+		return extensions.BlockReconcile(
+			fmt.Sprintf(
+				"Flexible Server is in provisioning state %q",
+				*server.Status.State)), nil
```

Comment:
Do we 100% know that `server.Status.State` is not nil? Otherwise we could panic here.

Reply:
Theoretically, `flexibleServerStateBlocksReconciliation()` ensures that `State` is not nil, but to hedge against future issues I refactored anyway.